### ISSUE TITLE
Best way to inject search query?

### ISSUE THREAD

**Issue by rtman** [algolia/react-instantsearch#862]
How should I inject a query into a search page?

My use case is this:

I have a tag selection page, when one tag is selected it brings up a searchResults screen with the relevant hits.

I've tried this:

```
export const UpdateQuery = connectSearchBox(({ refine, currentRefinement, query }) => {
  refine(query);
  return(
    null
  )
})
```
It works but it feels wrong, and it messes with subsequent searches.

The below works as an example, but I can't figure out how to use searchableAttributes instead of filters or filterFacets, since I don't want my tags to be a facet just a searchableAttribute.
```<Configure filters='companyName:Winehouse'/>```

The attribute I want to match to is an array containing various tags and I want to be able to specify a tag in the configure and get the resulting hits that contain that tag. So the array is stored like this in my index:
```tags = ['Wine', 'Red Wine']```

I tried the following configure but I'm assuming it's wrong:

```
this.query = 'Wine'
<Configure filters={'tags:${this.query}'}/>
```

Is `Configure` the best way to do this? If so how can I accomplish the above?

**Comment by samouss** [comment]
Hi @rtman,

I try to understand your use case. But why you want to use your tags as `searchableAttributes` rather than `facets` ? They can be both at the same time if you need it.

You can do it using `<Configure>` but it's probably easier to use a [Virtual Widget](https://community.algolia.com/react-instantsearch/guide/Virtual_widgets.html). It's a widget that will apply a refinement without rendering anything. In you case it will be a `VirtualMenu`, since it seems you want to refine only one value at a time.

```jsx
import { InstantSearch, SearchBox, Hits } from 'react-instantsearch/dom';
import { connectMenu } from 'react-instantsearch/connectors';

const VirtualMenu = connectMenu(() => null);

const App = () => (
  <InstantSearch
    appId="latency"
    apiKey="6be0576ff61c053d5f9a3225e2a90f76"
    indexName="bestbuy"
  >
    <VirtualMenu attributeName="category" defaultRefinement="iPad" />
    <SearchBox />

    <Hits />
  </InstantSearch>
);
```

**Comment by rtman** [comment]
> I try to understand your use case. But why you want to use your tags as searchableAttributes rather than facets ? They can be both at the same time if you need it.

I just dont need the tags to be available as refinements in the refinement list, i thought if you make them facets then it gets added to the refinement list

Great thanks I will try `VirtualMenu`

**Comment by rtman** [comment]
So the difference that I can see with `Configure` vs `VirtualMenu` is that `VirtualMenu` will update the currentRefinements widget and I don't necessarily want that as my searchResults screen already has the selected category or tag as the header.

An issue I've encountered with both of them is that I can't figure out how to set an element of an array.

The attribute I want to match is an element in an array containing various tags. I want to be able to specify a tag in the configure/Virtual widget and get the resulting hits that contain that tag. So the array is stored like this in my index:
tags = ['Wine', 'Red Wine']

Can you show me how this can be done?

**Comment by samouss** [comment]
> So the difference that I can see with Configure vs VirtualMenu is that VirtualMenu will update the currentRefinements widget and I don't want that as my searchResults screen already has the selected category or tag as the header and it shouldn't be removeable. Is there any way to disable that?

Yes you are right in that case you have two solutions: 

- The first one is to stick with the `VirtualMenu` but you will need to filter the elements that you want in your `CurrentRefinements`. You can achieve that with the [`transformItems`]() prop:

```jsx
import { InstantSearch, CurrentRefinements, Hits } from 'react-instantsearch/dom';
import { connectMenu } from 'react-instantsearch/connectors';

const VirtualMenu = connectMenu(() => null);

const App = () => (
  <InstantSearch
    appId="latency"
    apiKey="6be0576ff61c053d5f9a3225e2a90f76"
    indexName="bestbuy"
  >
    <VirtualMenu attributeName="category" defaultRefinement="PS4 Consoles" />

    <CurrentRefinements transformItems={items => items.filter(item => item.attributeName !== 'category')} />
    <Hits />
  </InstantSearch>
);
```

- The second approach is to use the [`Configure`](https://community.algolia.com/react-instantsearch/widgets/Configure.html) widget instead of the `VirtualMenu`. But you need to set the filters with the correct shape. It's one of the advantage of using the `Menu` as a virtual widget. You don't need to have the knowledge of the Algolia query parameters.  In your previous example it doesn't work when you have a space inside the variable `query`. You need to put double quotes around the value:

```jsx
import { InstantSearch, Configure, CurrentRefinements, Hits } from 'react-instantsearch/dom';

const App = () => (
  <InstantSearch
    appId="latency"
    apiKey="6be0576ff61c053d5f9a3225e2a90f76"
    indexName="bestbuy"
  >
    <Configure filters={'category:"PS4 Consoles"'} />

    <CurrentRefinements />
    <Hits />
  </InstantSearch>
);
```

**Comment by rtman** [comment]
Ah ok, I will look into it.

The issue is that I have a 3 drawer components (`Sort`, `refinements`, `hierarchicalMenu`) on the page and they are each running an instance of `<InstantSearch>`. Then there is the main instance displaying `currentRefinements`, `hits`, `breadCrumb` etc. I also have another for a modal which contains the searching widget. So I'm not quite sure how to keep them all in sync, as I think I need a new instance for each component (Drawer or modal).

Also I thought having the following props in each would keep them in sync?

```
                            onSearchStateChange={this.onSearchStateChange}
                            searchState={this.state.searchState}
```

Here is my render:

```
	render() {
		return (
            <Drawer 
                type="overlay"
                open={this.state.refineDrawerVisible}
                openDrawerOffset={0.5}
                closedDrawerOffset={0}
                //panCloseMask={0.6}
                //closedDrawerOffset={-3}
                onCloseStart={() => {this.setState({refineDrawerVisible: false});}} 
                tapToClose={true}
                side={'right'} 
                styles={drawerStyles}
                    tweenHandler={(ratio) => ({
                    main: { opacity:(2-ratio)/2 }
                })}
                content={
                    <ScrollView>
                        <InstantSearch
                            appId={algoliaAppId}
                            apiKey={algoliaApiKey}
                            indexName={currentIndex}
                            onSearchStateChange={this.onSearchStateChange}
                            searchState={this.state.searchState}>
                                <DrawerHeader headerTitle={'Refine'} drawerVisible={'refineDrawerVisible'}/>
                                <DrawerCategories
                                    onSearchStateChange={this.onSearchStateChange}
                                    searchState={this.state.searchState}
                                    indexName={currentIndex}
                                    attributeName={'companyName'}
                                    header={'Company'}/>
                                <Separator/>
                                <DrawerCategories
                                    onSearchStateChange={this.onSearchStateChange}
                                    searchState={this.state.searchState}
                                    indexName={currentIndex}
                                    attributeName={'price'}
                                    header={'Price'}/>
                                <Separator/>
                                <DrawerCategories
                                    onSearchStateChange={this.onSearchStateChange}
                                    searchState={this.state.searchState}
                                    indexName={currentIndex}
                                    attributeName={'unit'}
                                    header={'Unit'}/>
                                <Separator/>
                                <DrawerCategories
                                    onSearchStateChange={this.onSearchStateChange}
                                    searchState={this.state.searchState}
                                    indexName={currentIndex}
                                    attributeName={'range'}
                                    header={'Range'}/>
                        </InstantSearch>
                    </ScrollView>
                }>
            <Drawer 
                type="overlay"
                open={this.state.hierarchicalDrawerVisible}
                openDrawerOffset={0.5}
                closedDrawerOffset={0}
                //panCloseMask={0.6}
                //closedDrawerOffset={-3}
                onCloseStart={() => {this.setState({hierarchicalDrawerVisible: false});}} 
                tapToClose={true}
                side={'right'} 
                styles={drawerStyles}
                    tweenHandler={(ratio) => ({
                    main: { opacity:(2-ratio)/2 }
                })}
                content={
                    <ScrollView>
                        <InstantSearch
                            appId={algoliaAppId}
                            apiKey={algoliaApiKey}
                            indexName={currentIndex}
                            onSearchStateChange={this.onSearchStateChange}
                            searchState={this.state.searchState}>
                            <DrawerHeader headerTitle={'Categories'} drawerVisible={'hierarchicalDrawerVisible'}/>
                            <DrawerHierarchicalMenu
                                onSearchStateChange={this.onSearchStateChange}
                                searchState={this.state.searchState}
                                indexName={currentIndex}
                                attributes={categories}/>
                        </InstantSearch>
                    </ScrollView>
                }>
                <Drawer 
                    type="overlay"
                    open={this.state.sortDrawerVisible}
                    openDrawerOffset={0.5}
                    closedDrawerOffset={0}
                    //panCloseMask={0.6}
                    //closedDrawerOffset={-3}
                    onCloseStart={() => {this.setState({sortDrawerVisible: false});}}
                    tapToClose={true}
                    side={'right'} 
                    styles={drawerStyles}
                        tweenHandler={(ratio) => ({
                        main: { opacity:(2-ratio)/2 }
                    })}
                    content={
                        <ScrollView>
                            <InstantSearch
                                appId={algoliaAppId}
                                apiKey={algoliaApiKey}
                                indexName={currentIndex}
                                onSearchStateChange={this.onSearchStateChange}
                                searchState={this.state.searchState}>
                                <DrawerHeader headerTitle={'Sort'} drawerVisible={'sortDrawerVisible'}/>
                                <DrawerSortBy
                                        indexName={currentIndex}
                                        onSearchStateChange={this.onSearchStateChange}
                                        searchState={this.state.searchState}
                                        sortItems={sortItems}/>
                            </InstantSearch>
                        </ScrollView>
                    }>
        			<View style={styles.container}>
                        <InstantSearch
                            appId={algoliaAppId}
                            apiKey={algoliaApiKey}
                            indexName='tags'
                            onSearchStateChange={this.onSearchStateChange}
                            searchState={this.state.searchState}>
                            <AutoComplete 
                                indexName='tags'
                                modalVisible={this.state.modalVisible}
                                setModalVisible={this.setModalVisible}
                                onPressHit={this.onPressSearchHit}
                                fieldOne='tag'/>
                        </InstantSearch>
                        <InstantSearch
                            appId={algoliaAppId}
                            apiKey={algoliaApiKey}
                            indexName={currentIndex}
                            onSearchStateChange={this.onSearchStateChange}
                            searchState={this.state.searchState}
                            root={{
                              Root: View,
                              props: {
                                style: {
                                  flex: 1
                                }
                              }
                            }}>
                            <Configure filters={`${this.attributeName}:${this.query}`}/>
                            {/*VirtualMenu attributeName={this.attributeName} defaultRefinement={this.query}/>*/}
                            <View style={{flexDirection: 'row', alignItems: 'center', justifyContent: 'space-between', padding: 8}}>
                                <View style={{padding: 8}}>
                                    <TouchableOpacity style={{flexDirection: 'row', alignItems: 'center', justifyContent: 'center', backgroundColor: Colors.backgroundColor, padding: 0}} onPress={() => {this.setState({sortDrawerVisible: true});}}>
                                        <Icon name='sort' size={Styles.primaryIconSize} color={Colors.secondaryColor}/>
                                        <Text style={{color: Colors.secondaryColor}}>
                                            Sort
                                        </Text>
                                    </TouchableOpacity>
                                </View>
                                <View style={{padding: 8}}>
                                    <TouchableOpacity style={{flexDirection: 'row', alignItems: 'center', justifyContent: 'center', backgroundColor: Colors.backgroundColor, padding: 0}} onPress={() => {this.setState({refineDrawerVisible: true});}}>
                                        <Icon name='filter' size={Styles.primaryIconSize} color={Colors.secondaryColor}/>
                                        <Text style={{color: Colors.secondaryColor}}>
                                            Refine
                                        </Text>
                                    </TouchableOpacity>
                                </View>
                                <View style={{padding: 8}}>
                                    <TouchableOpacity style={{flexDirection: 'row', alignItems: 'center', justifyContent: 'center', backgroundColor: Colors.backgroundColor, padding: 0}} onPress={() => {this.setState({hierarchicalDrawerVisible: true})}}>
                                        <Icon name='format-list-bulleted' size={Styles.primaryIconSize} color={Colors.secondaryColor}/>
                                        <Text style={{color: Colors.secondaryColor}}>
                                            Categories
                                        </Text>
                                    </TouchableOpacity>
                                </View>
                                <View style={{padding: 8}}>
                                    <TouchableOpacity style={{flexDirection: 'row', alignItems: 'center', justifyContent: 'center', backgroundColor: Colors.backgroundColor, padding: 0}} onPress={() => {this.setState({mapDrawer: true});}}>
                                        <Icon name='google-maps' size={Styles.primaryIconSize} color={Colors.secondaryColor}/>
                                        <Text style={{color: Colors.secondaryColor}}>
                                            Map
                                        </Text>
                                    </TouchableOpacity>
                                </View>
                            </View>
                            <Separator/>
                            <CurrentRefinements/>
                            <Separator/>
                            <Breadcrumb
                                attributes={categories}
                                rootURL={null}
                                separator=' > '
                            />
                            <Hits 
                                fieldOne="itemName"
                                fieldTwo="price"
                                fieldThree="companyName"
                                uid={this.user.uid}
                                navigation={this.props.navigation}
                                usersRef={this.usersRef}
                                chatsRef={this.chatsRef}
                                onPressHit={this.onPressHit}/>
                            <VirtualRefinementList attributeName="companyName"/>
                            <VirtualRefinementList attributeName="price"/>
                            <VirtualRefinementList attributeName="unit"/>
                            <VirtualRefinementList attributeName="range"/>
                            <VirtualSortBy defaultRefinement={'items'} items={sortItems}/>
                            <VirtualHierarchicalMenu                                
                                attributes={categories}/>
                        </InstantSearch>
                    </View>
            </Drawer>
            </Drawer>
            </Drawer>
		);
	}
}
```

**Comment by rtman** [comment]
Actually would I not need to update the `searchState` on `componentWillMount` so that all the instances are in sync? Because I'm loading this component with category or tag selected as a 'search' and I think the instances don't know about it.

**Comment by samouss** [comment]
Actually in your case you don't have to use the `searchState` & `onSearchStateChange`. To avoid the synchronisation step you need to move the `<InstantSearch>` component higher in the hierarchy in order to allow all your widgets to have access to it. The widgets don't have to be direct children of the root component. In your example you will probably need to put it above the first drawer, then you can use the widgets inside your differents drawers without the need to recreate a new instance for each of them.

**Comment by rtman** [comment]
Wow ok, that worked great! I removed the `<InstantSearch>` + `searchState` + `onSearchStateChange` from the `refinementList`, `sortBy` and `hierarchicalMenu` widgets, and moved the `<InstantSearch>` instance to before the first drawer 👍 OH and all the virtual widgets were unnecessary after this.

Just a couple more things:

For the `hierarchicalMenu` the irrelevant options are removed, however the hierarchy isn't traversed (open to the correct level) when I view it. I suppose that has something to do with the way I've coded my component so I will investigate that.

My modal AutoComplete widget is running on another index and I can't get it to work with the current setup, before I was declaring a seperate `<InstantSearch>` instance for it since it's using a different index, is this correct? I also don't know where to put it, if I nest it it breaks things and react-native won't let me put it after the other instance without an enclosing `<View>` which seems to break everything.
